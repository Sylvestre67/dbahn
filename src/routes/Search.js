import React from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import debounce from 'lodash.debounce';
import { withStyles } from '@material-ui/core/styles';

import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';

import SearchResults from '../components/SearchResults';

const styles = theme => ({
	root: {},
	form: {},
	searchField: {
		width: '100%',
	},
});

export class Search extends React.Component {
	constructor(props) {
		super(props);

		this.handleSearch = this.handleSearch.bind(this);

		this.updateLocationQuery = this.updateLocationQuery.bind(this);
		this.updateLocationQuery = debounce(this.updateLocationQuery, 500);

		this.state = {
			search: '',
		};
	}

	componentDidMount() {
		const { location } = this.props;
		const { search } = location;

		if (search.length > 0) {
			this.setState({ search: search.replace('?q=', '') });
		}
	}

	handleSearch(e) {
		const { history, location } = this.props;
		this.setState({ search: e.target.value }, () => {
			this.updateLocationQuery();
		});
	}

	updateLocationQuery() {
		const { history, location } = this.props;

		history.push({
			pathname: location.pathname,
			search: `q=${this.state.search}`,
		});
	}

	render() {
		const { classes } = this.props;
		const { search } = this.state;

		return (
			<React.Fragment>
				<Grid container>
					<Grid item xs={12}>
						<Typography variant="h5">
							Search for a DB network station
						</Typography>
					</Grid>
				</Grid>

				<Grid container justify="center">
					<Grid item xs={12} sm={12} lg={6}>
						<form
							className={classes.form}
							noValidate
							autoComplete="off">
							<TextField
								id="station-search"
								label="Search"
								value={search}
								className={classes.searchField}
								onChange={e => {
									this.handleSearch(e);
								}}
								margin="normal"
							/>
						</form>
					</Grid>
				</Grid>

				<Grid container>
					<Grid item xs={12}>
						<SearchResults />
					</Grid>
				</Grid>
			</React.Fragment>
		);
	}
}

Search.propTypes = {
	classes: PropTypes.shape().isRequired,
};

export default withStyles(styles)(withRouter(Search));
