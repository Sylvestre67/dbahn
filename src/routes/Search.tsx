import React from 'react';
import { withRouter, RouteComponentProps } from 'react-router-dom';

import debounce from 'lodash.debounce';

import { withStyles, createStyles, WithStyles } from '@material-ui/core';

import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';

import SearchResults from '../components/SearchResults';

const styles = () =>
	createStyles({
		root: {},
		form: {},
		searchField: {
			width: '100%',
		},
	});

interface SearchStyle extends WithStyles<typeof styles> {}

interface SearchState {
	search: string;
}

export class Search extends React.Component<
	SearchStyle & RouteComponentProps,
	SearchState
> {
	constructor(props: SearchStyle & RouteComponentProps) {
		super(props);

		this.handleSearch = this.handleSearch.bind(this);

		this.updateLocationQuery = this.updateLocationQuery.bind(this);
		this.updateLocationQuery = debounce(this.updateLocationQuery, 500);

		this.state = {
			search: '',
		};
	}

	componentDidMount() {
		const { location } = this.props;
		const { search } = location;

		if (search.length > 0) {
			this.setState({ search: search.replace('?q=', '') });
		}
	}

	handleSearch(e: React.ChangeEvent<HTMLInputElement>) {
		this.setState({ search: e.target.value }, () => {
			this.updateLocationQuery();
		});
	}

	updateLocationQuery() {
		const { history, location } = this.props;

		history.push({
			pathname: location.pathname,
			search: `q=${this.state.search}`,
		});
	}

	render(): JSX.Element {
		const { classes } = this.props;
		const { search } = this.state;

		return (
			<React.Fragment>
				<Grid container>
					<Grid item xs={12}>
						<Typography variant="h5">
							Search for a DB network station
						</Typography>
					</Grid>
				</Grid>

				<Grid container justify="center">
					<Grid item xs={12} sm={12} lg={6}>
						<form
							className={classes.form}
							noValidate
							autoComplete="off">
							<TextField
								id="station-search"
								label="Search"
								value={search}
								className={classes.searchField}
								onChange={this.handleSearch}
								margin="normal"
							/>
						</form>
					</Grid>
				</Grid>

				<Grid container style={{ flex: '1 0 auto' }} justify="center">
					<Grid item xs={12} sm={12} lg={6}>
						<SearchResults />
					</Grid>
				</Grid>
			</React.Fragment>
		);
	}
}

export default withStyles(styles)(withRouter(Search));
