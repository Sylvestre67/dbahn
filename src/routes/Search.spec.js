import React from 'react';
import { shallow } from 'enzyme';

import Search from './Search';
let component;

const props = {
	classes: { root: 'testing' },
};

describe('Search', () => {
	beforeAll(() => {
		component = shallow(<Search {...props} />);
	});

	it('should render correctly', () => {
		expect(component).toMatchSnapshot();
	});
});
