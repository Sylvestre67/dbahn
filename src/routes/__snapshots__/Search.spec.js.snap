// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render correctly 1`] = `
<withRouter(Search)
  classes={
    Object {
      "form": "withRouter-Search--form-2",
      "root": "withRouter-Search--root-1 testing",
      "searchField": "withRouter-Search--searchField-3",
    }
  }
/>
`;
